### PR TITLE
Updates the camera for webots so i can test behavior 

### PR DESCRIPTION
**Pull request by tayloryoung6396**

### CODE REVIEW

**Review by tayloryoung6396 — COMMENTED** [review]
Hcw is incorrect atm

File: module/platform/Webots/data/config/camera.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# serial_number: 0
+
+lens:
+  # Lens:
```
Comment (ysims) [inline module/platform/Webots/data/config/camera.yaml:4]: What are all these comments for

File: module/platform/Webots/data/config/camera.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  # NOTE: all lens parameters are based on a full size image (without width/height/offset values)
+
+  # Lens projection type
+  projection: RECTILINEAR
+  # Normalised focal length: focal length in pixels / image width
+  focal_length: 0.34767878619271286
+  # Normalised image centre offset: pixels from centre to optical axis / image width
+  centre: [0, 0]
+  # The polynomial distortion coefficients for the lens
+  k: [0.46673390301927425, 0.15017118169604132]
```
Comment (ysims) [inline module/platform/Webots/data/config/camera.yaml:21]: I thought this was meant to be [0,0]

File: module/platform/Webots/src/Webots.cpp [hunk 3]
```diff
@@ -695,7 +764,85 @@ namespace module::platform {
...
+            image->lens = context->lens;
+            image->Hcw  = Hcw.matrix();
+
             emit(image);
         }
     }
 }  // namespace module::platform
+
+
+// message Image {
```
Comment (ysims) [inline module/platform/Webots/src/Webots.cpp:818]: What is all this?

File: module/support/NUsight/data/config/NUsight.yaml [hunk 4]
```diff
@@ -12,21 +12,21 @@ reaction_handles:
   # Overview information for the dashboard, designed for minimum bandwidth
   overview: true
   # Motor positions, orientation, etc.
-  sensor_data: false
+  sensor_data: true
```
Comment (ysims) [inline module/support/NUsight/data/config/NUsight.yaml:15]: Revert changes to this file

**Review by ysims — CHANGES_REQUESTED** [review]
Extra changes need to be removed.

File: module/platform/Webots/src/Webots.cpp (hunk 3 above)
Comment (tayloryoung6396) [reply module/platform/Webots/src/Webots.cpp:818]: Notes for myself

File: module/platform/Webots/data/config/camera.yaml (hunk 2 above)
Comment (tayloryoung6396) [reply module/platform/Webots/data/config/camera.yaml:21]: It is in another commit

File: module/platform/Webots/data/config/camera.yaml (hunk 1 above)
Comment (tayloryoung6396) [reply module/platform/Webots/data/config/camera.yaml:4]: Will be populated when i know what they are

File: module/support/NUsight/data/config/NUsight.yaml (hunk 4 above)
Comment (tayloryoung6396) [reply module/support/NUsight/data/config/NUsight.yaml:15]: Will when I'm done

**Review by tayloryoung6396 — COMMENTED** [review]
This now works, however odometry breaks it, if you remove the odometry updates (and stop it walking) you can see it working

**Comment by tayloryoung6396** [comment]
Screenshot from 2021-06-17 10-29-22

File: module/platform/Webots/data/config/camera.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    - [0.01728764296000275, 0.9986932814319833, 0.04809227613146373, 0.03529909622045881]
+    - [0.014843552535558163, -0.04835051478781678, 0.9987201292902445, 0.0713674563254241]
+    - [0.0, 0.0, 0.0, 1.0]
+
+settings:
+  # PixelFormat: BayerRG8
+  Width: 640
+  Height: 480
+  OffsetX: 0
+  OffsetY: 0
```
Comment (wongjoel) [inline module/platform/Webots/data/config/camera.yaml:36]: Verified that this matches what we set in Webots for data collection (Assuming that Hpc = Hxl)

**Comment by wongjoel** [comment]
Not sure I understand exactly what's happening in `webots.cpp` in the `on<Configuration>("camera.yaml").then([this](const Configuration& config)`, but otherwise it looks fine to me.

**Comment by tayloryoung6396** [comment]
> Not sure I understand exactly what's happening in `webots.cpp` in the `on<Configuration>("camera.yaml").then([this](const Configuration& config){...}` section, but otherwise it looks fine to me.

**Comment by tayloryoung6396** [comment]
> Not sure I understand exactly what's happening in `webots.cpp` in the `on<Configuration>("camera.yaml").then([this](const Configuration& config){...}` section, but otherwise it looks fine to me.

Basically setting up the parameters of the camera that are required elsewhere, this is done the came as Camera.cpp. It could be simplified, but if we ever want to change the camera in webots or add additional cameras i think it will be needed

File: module/platform/Webots/src/Webots.cpp [hunk 6]
```diff
@@ -242,6 +247,86 @@ namespace module::platform {
...
+            // auto it = cameras
+            //               .insert(std::make_pair(serial_number,
+            //                                      CameraContext{
+            //                                          *this,
+            //                                          name,
+            //                                          num_cameras++,
+            //                                          Image::Lens(),      // Lens is constructed in settings
+            //                                          Eigen::Affine3d(),  // Hpc is set in settings
+            //                                      }))
+            //               .first;
```
Comment (wongjoel) [inline module/platform/Webots/src/Webots.cpp:266]:
```suggestion
            // Strip the .yaml off the name of the file to get the name of the camera
            std::string name = ::basename(config.fileName.substr(0, config.fileName.find_last_of('.')).c_str());
```

Should probably remove the commented out code?

Comment (tayloryoung6396) [reply module/platform/Webots/src/Webots.cpp:266]: Seems like a good idea

**Review by wongjoel — APPROVED** [review]
Given the code I don't fully understand was basically copied in from `camera.cpp`, I'm happy with this.

**Pull request merged**